src/roadmap/v1/upcoming.py

```diff
@@ -38,37 +42,87 @@ class Architecture(StrEnum):
     none = ""
 
 
-class UpcomingDetails(BaseModel):
+class UpcomingInputDetails(BaseModel):
     architecture: Architecture | None
     detailFormat: int
     summary: str
-    potentiallyAffectedSystems: int
     trainingTicket: str
     dateAdded: date = Field(default_factory=date.today)
     lastModified: Date
 
 
-class Upcoming(BaseModel):
+class UpcomingInput(BaseModel):
     name: str
     type: UpcomingType
+    package: str
     release: str
     date: Date
-    details: UpcomingDetails
+    details: UpcomingInputDetails
+
+
+class UpcomingOutputDetails(BaseModel):
+    architecture: Architecture | None
+    detailFormat: int
+    summary: str
+    trainingTicket: str
+    dateAdded: date = Field(default_factory=date.today)
+    lastModified: Date
+    potentiallyAffectedSystemsCount: int
+    potentiallyAffectedSystems: list[UUID]
+
+
+class UpcomingOutput(BaseModel):
+    name: str
+    type: UpcomingType
+    release: str
+    date: Date
+    details: UpcomingOutputDetails
 
 
 class WrappedUpcoming(BaseModel):
     meta: Meta
-    data: list[Upcoming]
+    data: list[UpcomingOutput]
 
 
 @lru_cache
-def read_upcoming_file(file: str | Path) -> list[Upcoming]:
-    with open(file, "r") as file:
-        return TypeAdapter(list[Upcoming]).validate_json(file.read())
-
-
-def get_upcoming_data(settings: t.Annotated[Settings, Depends(Settings.create)]) -> list[Upcoming]:
-    return read_upcoming_file(settings.upcoming_json_path)
+def read_upcoming_file(file: Path) -> list[UpcomingInput]:
+    return TypeAdapter(list[UpcomingInput]).validate_json(file.read_text())
+
+
+def get_upcoming_data(
+    systems_by_stream: t.Annotated[dict[AppStreamKey, list[UUID]], Depends(systems_by_app_stream)],
+    settings: t.Annotated[Settings, Depends(Settings.create)],
+) -> list[UpcomingOutput]:
+    keys_by_name = defaultdict(list)
+    [keys_by_name[a.name].append(a) for a in systems_by_stream.keys()]
+
+    result = []
+    for upcoming in read_upcoming_file(settings.upcoming_json_path):
+        systems = set()
+        for key in keys_by_name[upcoming.package]:
+            systems.update(systems_by_stream[key])
+
+        details = UpcomingOutputDetails(
+            architecture=upcoming.details.architecture,
+            detailFormat=upcoming.details.detailFormat,
+            summary=upcoming.details.summary,
+            trainingTicket=upcoming.details.trainingTicket,
+            dateAdded=upcoming.details.dateAdded,
+            lastModified=upcoming.details.lastModified,
```

Comment:
This looks like `upcoming.details` could just be passed and `UpcomingOutput` should do something with it , or just reuse the existing values instead of typing them all out.
```suggestion
            **upcoming.details.model_dump(),
```

Reply:
line 112 and 113 add more data. I didn't know a fancier pattern for doing what I was trying to do here.